src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,262 @@
+    m_COPlaying.forceSet(status == Status::Playing ? 1 : 0);
+    // TODO(xerus) add blinking for Status::Recording & Status::Playing
+    //m_COIndicator.forceSet(status > Status::Empty ? 1 : 0);
+}
+
+MacroPointer MacroControl::getMacro() const {
+    return m_pMacro;
+}
+
+bool MacroControl::isRecording() const {
+    return getStatus() == Status::Armed || getStatus() == Status::Recording;
+}
+
+void MacroControl::play() {
+    DEBUG_ASSERT(m_pMacro);
+    m_iNextAction = 1;
+    setStatus(Status::Playing);
+    m_pMacro->setState(Macro::StateFlag::Enabled);
+}
+
+void MacroControl::stop() {
+    DEBUG_ASSERT(m_pMacro);
+    m_iNextAction = INT_MAX;
+    setStatus(Status::Recorded);
+    m_pMacro->setState(Macro::StateFlag::Enabled, false);
+}
+
+void MacroControl::updateRecording() {
+    //qCDebug(macroLoggingCategory) << QThread::currentThread() << QTime::currentTime() << "Update recording status:" << getStatus() << "recording:" << isRecording();
+    VERIFY_OR_DEBUG_ASSERT(isRecording()) {
+        return;
+    }
+    bool actionsRecorded = false;
+    while (MacroAction* action = m_recordedActions.front()) {
+        m_pMacro->addAction(*action);
+        m_recordedActions.pop();
+        actionsRecorded = true;
+    }
+    if (actionsRecorded && getStatus() == Armed) {
+        setStatus(Recording);
+    }
+}
+
+void MacroControl::stopRecording() {
+    VERIFY_OR_DEBUG_ASSERT(isRecording()) {
+        return;
+    }
+    m_updateRecordingTimer.stop();
+    updateRecording();
+    if (getStatus() == Status::Armed) {
+        setStatus(Status::Empty);
+        if (m_pMacro->getLabel() == "[Recording]") {
+            m_pMacro->setLabel("");
+        }
+    } else {
+        // This will still be the position of the previous track when called from trackLoaded
+        // since trackLoaded is invoked before the SampleOfTrack of the controls is updated.
+        m_pMacro->setEnd(getSampleOfTrack().current / mixxx::kEngineChannelCount);
+        if (m_pMacro->getLabel() == "[Recording]") {
+            // Automatically set the start position in seconds as label if there
+            // is no user-defined one
+            int secPos = m_pMacro->getStartSamplePos() / getSampleOfTrack().rate;
+            m_pMacro->setLabel(QString::number(secPos, 'f', 1));
+        }
+        setStatus(Status::Recorded);
+        if (m_pMacro->isEnabled()) {
+            slotGotoPlay();
+        }
+    }
+}
+
+void MacroControl::slotRecord(double value) {
+    if (!value)
+        return;
+    if (getStatus() == Status::Empty) {
+        setStatus(Status::Armed);
+        DEBUG_ASSERT(m_updateRecordingTimer.thread() == QThread::currentThread());
+        m_updateRecordingTimer.start(kRecordingTimerInterval);
+        if (m_pMacro->getLabel().isEmpty()) {
+            m_pMacro->setLabel("[Recording]");
+        }
+    } else if (isRecording()) {
+        stopRecording();
+    }
+}
+
+void MacroControl::slotToggle(double value) {
+    if (!value)
```

Comment:
missing braces

Reply:
you don't need to comment such things 4 times, I always have a look around after these things :)